source/common/upstream/load_balancer_impl.cc

```diff
@@ -46,48 +59,92 @@ bool LoadBalancerBase::isGlobalPanic() {
   return false;
 }
 
-const std::vector<HostPtr>& LoadBalancerBase::hostsToUse() {
-  ASSERT(host_set_.healthyHosts().size() <= host_set_.hosts().size());
+std::vector<uint64_t>
+LoadBalancerBase::calculateZonePercentage(const std::vector<std::vector<HostPtr>>& hosts_per_zone) {
+  std::vector<uint64_t> percentage(hosts_per_zone.size());
 
-  if (host_set_.hosts().empty() || isGlobalPanic()) {
-    return host_set_.hosts();
+  uint64_t total_hosts = 0;
+  for (const auto& zone_hosts : hosts_per_zone) {
+    total_hosts += zone_hosts.size();
   }
 
-  if (earlyExitNonZoneRouting()) {
-    return host_set_.healthyHosts();
+  if (total_hosts != 0) {
+    size_t pos = 0;
+    for (const auto& zone_hosts : hosts_per_zone) {
+      percentage[pos++] = 10000ULL * zone_hosts.size() / total_hosts;
+    }
   }
 
+  return percentage;
+}
+
+const std::vector<HostPtr>& LoadBalancerBase::tryChooseLocalZoneHosts() {
   // At this point it's guaranteed to be at least 2 zones.
-  uint32_t number_of_zones = host_set_.healthyHostsPerZone().size();
-  ASSERT(number_of_zones >= 2U);
-  const std::vector<HostPtr>& local_zone_healthy_hosts = host_set_.healthyHostsPerZone()[0];
+  ASSERT(host_set_.healthyHostsPerZone().size() >= 2U);
 
-  // If number of hosts in a local zone big enough then route all requests to the same zone.
-  if (local_zone_healthy_hosts.size() * number_of_zones >= host_set_.healthyHosts().size()) {
+  std::vector<uint64_t> local_percentage =
```

Comment:
Or just do the pre-calculation work now. I would actually prefer you just do that.